src/platform/Darwin/DnssdContexts.cpp

```diff
@@ -0,0 +1,408 @@
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR MdnsContexts::GetRegisterContextOfType(const char * type, GenericContext ** context)
+{
+    bool found = false;
+    std::vector<GenericContext *>::iterator iter;
+
+    for (iter = mContexts.begin(); iter != mContexts.end(); iter++)
+    {
+        if ((*iter)->type == ContextType::Register && ((RegisterContext *) (*iter))->matches(type))
+        {
+            *context = *iter;
+            found    = true;
+            break;
+        }
+    }
+
+    return found ? CHIP_NO_ERROR : CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+RegisterContext::RegisterContext(const char * sType, DnssdPublishCallback cb, void * cbContext)
+{
+    type     = ContextType::Register;
+    context  = cbContext;
+    callback = cb;
+
+    Platform::CopyString(mType, sType);
+}
+
+void RegisterContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Register (%s)", Error::ToString(err));
+    callback(context, nullptr, CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void RegisterContext::DispatchSuccess()
+{
+    callback(context, mType, CHIP_NO_ERROR);
+}
+
+BrowseContext::BrowseContext(void * cbContext, DnssdBrowseCallback cb, DnssdServiceProtocol cbContextProtocol)
+{
+    type     = ContextType::Browse;
+    context  = cbContext;
+    callback = cb;
+    protocol = cbContextProtocol;
+}
+
+void BrowseContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Browse (%s)", Error::ToString(err));
+    callback(context, nullptr, 0, CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void BrowseContext::DispatchSuccess()
+{
+    callback(context, services.data(), services.size(), CHIP_NO_ERROR);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+ResolveContext::ResolveContext(void * cbContext, DnssdResolveCallback cb, chip::Inet::IPAddressType cbAddressType)
+{
+    type     = ContextType::Resolve;
+    context  = cbContext;
+    callback = cb;
+    protocol = GetProtocol(cbAddressType);
+}
+
+ResolveContext::~ResolveContext()
+{
+    RemoveInterfaces();
+}
+
+void ResolveContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Resolve (%s)", Error::ToString(err));
+    callback(context, nullptr, Span<Inet::IPAddress>(), CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void ResolveContext::DispatchSuccess()
+{
+    for (auto & service : services)
+    {
+        auto & ips = addresses[service.first];
+
+        // Some interface may not have any ips, just ignore them.
+        if (ips.size() == 0)
+        {
+            continue;
+        }
+
+        // Use the first IP we got for the DnssdService.
+        service.second.mAddress.SetValue(ips.front());
+        ips.erase(ips.begin());
+        callback(context, &service.second, Span<Inet::IPAddress>(ips.data(), ips.size()), CHIP_NO_ERROR);
+        MdnsContexts::GetInstance().Remove(this);
+        return;
```

Comment:
I think this might be clearer as a `break` from the loop and a `MdnsContexts::GetInstance().Remove(this);` outside the loop, to make it very clear that we always `Remove()`.